### PR TITLE
Adaption of kernel optimizer

### LINKED ISSUE #274: Adapatation of KernelOptimizer to squlearn standards
The current implementation of the Kernel Optimization is somewhat misaligned with squlearn's design principles.

It would be beneficial to modify the implementation so that the KernelOptimizer is based on the KernelMatrixBase class. The inputs should include a KernelMatrix (KernelMatrixBase – currently either Fidelity or Projected Kernel), the loss function, and the optimizer. The optimization should be automatically triggered when the fit function in the high-level Kernel implementation is called.

### PR DESCRIPTION
**Pull request by DennisKleinhans**
- The KernelOptimizer extends the KernelMatrixBase so that it can be used as a kernel in the high-level methods. 
- The KernelOptimizer now accepts a KernelMatrix, a loss function and an optimizer. 
- The KernelOptimizer automatically optimizes the provided KernelMatrix when the fit method is called by checking the is_trainable property. 

closes #274

### CODE REVIEW

File: src/squlearn/kernel/matrix/kernel_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+        # Perform kernel optimization
+        loss_function = partial(self._loss.compute, data=X, labels=y)
+        opt_result = self._optimizer.minimize(fun=loss_function, x0=self._initial_parameters)
+        self._optimal_value = opt_result.fun
+        self._optimal_point = opt_result.x
+        self._opt_result = opt_result
+
+        # Assign optimal parameters to the quantum kernel
+        self.assign_parameters(self._opt_result.x)
```
Comment (David-Kreplin) [inline src/squlearn/kernel/matrix/kernel_optimizer.py:75]: Is this working? Somewhere I'm missing the call to ``self._quantum_kernel.assign_parameters()``

File: src/squlearn/kernel/matrix/kernel_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        )
+        self._quantum_kernel = quantum_kernel
+        self._loss = loss
+        self._optimizer = optimizer
+        self._initial_parameters = (
+            initial_parameters
+            if initial_parameters is not None
+            else self._quantum_kernel.parameters
+        )
+        self._opt_result = None
```
Comment (David-Kreplin) [inline src/squlearn/kernel/matrix/kernel_optimizer.py:46]: I think opt_result is not really needed

File: src/squlearn/kernel/matrix/kernel_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        self._optimal_value = None
+        self._optimal_point = None
```
Comment (David-Kreplin) [inline src/squlearn/kernel/matrix/kernel_optimizer.py:48]: I think a getter for the optimal_point would be great, I would also suggest to rename it into _optimal_parameters

File: src/squlearn/kernel/matrix/kernel_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        loss_function = partial(self._loss.compute, data=X, labels=y)
+        opt_result = self._optimizer.minimize(fun=loss_function, x0=self._initial_parameters)
+        self._optimal_value = opt_result.fun
+        self._optimal_point = opt_result.x
+        self._opt_result = opt_result
+
+        # Assign optimal parameters to the quantum kernel
+        self.assign_parameters(self._opt_result.x)
+
+        return self._opt_result
```
Comment (David-Kreplin) [inline src/squlearn/kernel/matrix/kernel_optimizer.py:77]: just return opt_result without storring it.

File: src/squlearn/kernel/matrix/kernel_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        self._quantum_kernel = quantum_kernel
+        self._loss = loss
+        self._optimizer = optimizer
+        self._initial_parameters = (
+            initial_parameters
+            if initial_parameters is not None
+            else self._quantum_kernel.parameters
+        )
+        self._opt_result = None
+        self._optimal_value = None
```
Comment (David-Kreplin) [inline src/squlearn/kernel/matrix/kernel_optimizer.py:47]: Probably not really needed.

**Review by David-Kreplin — CHANGES_REQUESTED** [review]
Great work so far! I've added only minor remarks. 
Could you also please add a small testjob for the kernel optimizer

File: src/squlearn/kernel/matrix/kernel_optimizer.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                "Quantum kernel cannot be fit because there are no training parameters specified."
+            )
+
+        # Perform kernel optimization
+        loss_function = partial(self._loss.compute, data=X, labels=y)
+        opt_result = self._optimizer.minimize(fun=loss_function, x0=self._initial_parameters)
+        self._optimal_parameters = opt_result.x
+
+        # Assign optimal parameters to the quantum kernel
+        self._quantum_kernel.assign_parameters(self._optimal_parameters)
```
Comment (David-Kreplin) [inline src/squlearn/kernel/matrix/kernel_optimizer.py:71]: Sorry to comment again, but there is missing a assign_parameters for the class (just call self._quantum_kernel.assign_parameters)

**Comment by David-Kreplin** [comment]
I did some adjustments, maybe you can check them @MoritzWillmann

**Review by David-Kreplin — APPROVED** [review]

**Comment by David-Kreplin** [comment]
I’ll take it from here. Most of the work looks great, but there are a few specifics that are a bit tedious to explain and fix, such as regularization.
